newton/_src/sim/collide_unified.py

```diff
@@ -0,0 +1,1197 @@
+        potentially colliding pairs are computed dynamically based on bounding sphere overlaps.
+
+        Args:
+            model (Model): The simulation model.
+            state (State): The current simulation state.
+
+        Returns:
+            Contacts: The generated contacts for the current state.
+        """
+        # Allocate new contact memory for contacts if needed (e.g., for gradients)
+        if self.contacts is None or self.requires_grad:
+            self.contacts = Contacts(
+                self.rigid_contact_max,
+                self.soft_contact_max,
+                requires_grad=self.requires_grad,
+                device=model.device,
+            )
+        else:
+            self.contacts.clear()
+
+        # output contacts buffer
+        contacts = self.contacts
+
+        # Clear counters at start of frame
+        self.broad_phase_pair_count.zero_()
+
+        # Compute AABBs for all shapes in world space (needed for both NXN and SAP)
+        # AABBs are computed using support function for most shapes
+        wp.launch(
+            kernel=compute_shape_aabbs,
+            dim=model.shape_count,
+            inputs=[
+                state.body_q,
+                model.shape_transform,
+                model.shape_body,
+                model.shape_type,
+                model.shape_scale,
+                model.shape_collision_radius,
+                model.shape_source_ptr,
+                self.rigid_contact_margin,
+            ],
+            outputs=[
+                self.shape_aabb_lower,
+                self.shape_aabb_upper,
+            ],
+            device=model.device,
+        )
+
+        # debug_a = model.shape_collision_group.numpy()
+        # debug_b = model.shape_group.numpy()
+
+        # Run appropriate broad phase
+        if self.broad_phase_mode == BroadPhaseMode.NXN:
+            # Use NxN all-pairs broad phase with AABB overlaps
+            self.nxn_broadphase.launch(
+                self.shape_aabb_lower,
+                self.shape_aabb_upper,
+                model.shape_thickness,  # Use thickness as cutoff
+                model.shape_collision_group,  # Collision groups for filtering
+                model.shape_group,  # Environment groups for filtering
+                model.shape_count,
+                self.broad_phase_shape_pairs,
+                self.broad_phase_pair_count,
+                device=model.device,
+            )
+        elif self.broad_phase_mode == BroadPhaseMode.SAP:
+            # Use sweep-and-prune broad phase with AABB sorting
+            self.sap_broadphase.launch(
+                self.shape_aabb_lower,
+                self.shape_aabb_upper,
+                model.shape_thickness,  # Use thickness as cutoff
+                model.shape_collision_group,  # Collision groups for filtering
+                model.shape_group,  # Environment groups for filtering
```

Comment:
Make sure both the NXN and SAP branches are unit-tested. `Model.shape_group` is now `Model.shape_world`.